File: src/javascript/algorithms/stack/balancedParentheses/balancedParentheses.js
/**
 * Balanced Parentheses (Valid Parentheses - LeetCode 20)
 *
 * Given a string containing brackets, determine if it is valid.
 * Approach 1 handles {}()[], Approach 2 also handles <>.
 *
 * Approach 1: Stack              TC: O(n)  SC: O(n)
 * Approach 2: HashMap + Stack    TC: O(n)  SC: O(n)
 */

// Approach 1: Stack
function validParenthesis1(s) {
    if (s.length % 2 !== 0) return false;
    const stack = [];
    for (const ch of s) {
        if ('{(['.includes(ch)) {
            stack.push(ch);
        } else {
            if (!stack.length) return false;
            const top = stack[stack.length - 1];
            if (ch === '}' && top !== '{') return false;
            if (ch === ')' && top !== '(') return false;
            if (ch === ']' && top !== '[') return false;
            stack.pop();
        }
    }
    return stack.length === 0;
}

// Approach 2: HashMap + Stack (also handles <>)
function validParenthesis2(s) {
    if (s.length % 2 !== 0) return false;
    const map = {')': '(', '}': '{', ']': '[', '>': '<'};
    const stack = [];
    for (const ch of s) {
        if (map[ch] !== undefined) {
            if (!stack.length || stack.pop() !== map[ch]) return false;
        } else {
            stack.push(ch);
        }
    }
    return stack.length === 0;
}

const commonCases = [
    { s: "{([])}", expected: true,  label: "{([])}" },
    { s: "{(][)}", expected: false, label: "{(][)}" },
    { s: "()",     expected: true,  label: "()" },
    { s: "()[]{}", expected: true,  label: "()[]{}" },
    { s: "(]",     expected: false, label: "(]" },
    { s: "([)]",   expected: false, label: "([)]" },
    { s: "{[]}",   expected: true,  label: "{[]}" },
    { s: "(())",   expected: true,  label: "(())" },
    { s: "(((", expected: false, label: "(((" },
];
const hmExtras = [
    { s: "<>",   expected: true,  label: "<>" },
    { s: "<(>)", expected: false, label: "<(>)" },
];

console.log("=== Approach 1: Stack ===");
commonCases.forEach(({ s, expected, label }) => {
    const result = validParenthesis1(s);
    console.log(`  [${result === expected ? "PASS" : "FAIL"}] "${label}": ${result}  (expected ${expected})`);
});

console.log("=== Approach 2: HashMap ===");
[...commonCases, ...hmExtras].forEach(({ s, expected, label }) => {
    const result = validParenthesis2(s);
    console.log(`  [${result === expected ? "PASS" : "FAIL"}] "${label}": ${result}  (expected ${expected})`);
});

File: src/python/algorithms/stack/balancedParentheses/balanced_parentheses.py
"""
Balanced Parentheses (Valid Parentheses - LeetCode 20)

Given a string containing brackets, determine if it is valid.
Approach 1 handles {}()[], Approach 2 also handles <>.

Approach 1: Stack              TC: O(n)  SC: O(n)
Approach 2: HashMap + Stack    TC: O(n)  SC: O(n)
"""


# Approach 1: Stack
def valid_parenthesis1(s):
    if len(s) % 2 != 0:
        return False
    stack = []
    for ch in s:
        if ch in '{([':
            stack.append(ch)
        else:
            if not stack: return False
            top = stack[-1]
            if ch == '}' and top != '{': return False
            if ch == ')' and top != '(': return False
            if ch == ']' and top != '[': return False
            stack.pop()
    return len(stack) == 0


# Approach 2: HashMap + Stack (also handles <>)
def valid_parenthesis2(s):
    if len(s) % 2 != 0:
        return False
    mapping = {')': '(', '}': '{', ']': '[', '>': '<'}
    stack = []
    for ch in s:
        if ch in mapping:
            if not stack or stack.pop() != mapping[ch]:
                return False
        else:
            stack.append(ch)
    return len(stack) == 0


if __name__ == "__main__":
    common_cases = [
        ("{([])}", True,  "{([])}"),
        ("{(][)}", False, "{(][)}"),
        ("()",     True,  "()"),
        ("()[]{}", True,  "()[]{}"),
        ("(]",     False, "(]"),
        ("([)]",   False, "([)]"),
        ("{[]}",   True,  "{[]}"),
        ("(())",   True,  "(())"),
        ("(((", False,  "((("),
    ]
    hm_extras = [
        ("<>",    True,  "<>"),
        ("<(>)",  False, "<(>)"),
    ]

    print("=== Approach 1: Stack ===")
    for s, expected, label in common_cases:
        result = valid_parenthesis1(s)
        status = "PASS" if result == expected else "FAIL"
        print(f'  [{status}] "{label}": {result}  (expected {expected})')

    print("=== Approach 2: HashMap ===")
    for s, expected, label in common_cases + hm_extras:
        result = valid_parenthesis2(s)
        status = "PASS" if result == expected else "FAIL"
        print(f'  [{status}] "{label}": {result}  (expected {expected})')
